### ISSUE TITLE
Better messages for unexpected exits

### ISSUE THREAD

**Issue by devinamatthews** [solomonik/ctf#15]
It would be great if there was some message indicated why CTF is exiting abnormally, something like:

* "Out of memory" (or "posix_memalign returned XXX")
* "Assertion failed on line XXX of YYY"

This would make tracking down issues much easier and help to distinguish from genuine segfaults. Also, wouldn't it be better to use MPI_Abort rather than dereference NULL?

**Comment by solomonik** [comment]
I dereference NULL in asserts when -DDEBUG is on, because this allows one
to get a trace via valgrind (or gdb) rather than when the application exits
normally or with MPI abort. This is how I always debug and what I would
also recommend using valgrind for you. I was not able to find a portable
solution for printing a stack trace manually.

When -DDEBUG is not on, when CTF detects an error it does some printf with
info and exits gracefully, but there are not nearly as many checks for
performance reasons.

**Comment by devinamatthews** [comment]
However it exits, something like ASSERT(condition, message) could print 
the message on failure (and with a macro wrapper automatically give the 
file and line). For example, I use the following macro in one of my codes:

```C
#define ALWAYS_ASSERT(cond,...) \
do { \
     if (!(cond)) \
     { \
         MASTER \
         { \
             fprintf(stderr, "%s(%d): assertion failed: ", __FILE__, 
__LINE__); \
             if (strcmp("()", STRINGIFY((__VA_ARGS__))) == 0) \
                 fprintf(stderr, STRINGIFY(cond)); \
             else \
                 fprintf(stderr, "" __VA_ARGS__); \
             fprintf(stderr, "\n"); \
         } \
         abort(); \
     } \
} while (0)
```

which support both ASSERT(cond) and ASSERT(cond,message) styles (note 
that MASTER is another macro which checks rank==0).

**Comment by devinamatthews** [comment]
Apparently Markdown doesn't work from e-mail...

**Comment by solomonik** [comment]
Sure I'll add a printf, its just that 99% of the time I am debugging I want
the full trace anyway, where this information would become redundant. The
location of the assert by itself is rarely enough info to debug anything.

**Comment by devinamatthews** [comment]
It helps a lot when you're not running in a debugger the first time or 
when the stack is corrupted. Also, I think a lot of these asserts could 
be left on without much (any?) performance penalty. Doing so, or having 
a mode to turn on assert but not DEBUG printf's would be helpful in 
catching errors like I had recently.

**Comment by jeffhammond** [comment]
That is correct. You can't even clean it up later (I've tried).

**Comment by jeffhammond** [comment]
We should aspire to support backtrack under Linux at least. I think it's feasible. Will investigate later. 

Blue Gene doesn't support Valgrind.

**Comment by devinamatthews** [comment]
On that note, libunwind is a great tool for this, but is there any library equivalent of addr2line (beyond calling it in a shell)? libbfd looks crazy hard to use.

**Comment by jeffhammond** [comment]
We should create a new issue for backtrace support and leave it as a long-term, low-priority feature request.  I'll try to look into it some day, but I suspect I will have three kids before it happens ;-)

**Comment by solomonik** [comment]
I have the below stack trace print thing,

https://github.com/solomonik/ctf/blob/master/src/interface/common.cxx#L156

which used to get called in ASSERT, but I stopped using it, because it does
not work anywhere except simple linux boxes where you can always run
valgrind anyway.

Also at the moment, CTF does not really support BG/Q anymore.

**Comment by solomonik** [comment]
Ok I reactivated the stack trace printing as well as added a print message that shows what and where the ASSERT is.  However, this info is only printed on processor 0, which means even if all processors fail there is no guarantee that the error will be printed.  This is one of the reasons I had taken this stuff out some time ago.  Printing errors on all processors becomes problematic when you start running on a few thousand of them.

Generally -DDEBUG and ASSERT() functionality is designed for people changing CTF and wanting to debug the resulting code, not for users of CTF.  Indeed when testing new functionality on top of CTF, they can help detect CTF bugs, but I think cases like the highly intricate error you ran into when running CCSDTQ in parallel are quite exceptional.
